src/startup.ts

```diff
@@ -1,60 +1,106 @@
+	setRuntimeConfig,
+} from "./runtimeConfig.js";
+import { inspect } from "util";
+import { stat, access } from "fs/promises";
+import { constants } from "fs";
+
+/**
+ * validates existence, permission, and that a path is a directory
+ * @param path string of path to validate
+ * @param optionName name of the configuration key
+ * @returns true if path exists and has required permission
+ */
+async function verifyPath(
+	path: string,
+	optionName: string,
+	permissions: number
+): Promise<boolean> {
+	try {
+		if ((await stat(path)).isDirectory()) {
+			await access(path, permissions);
+			return true;
+		}
+	} catch (error) {
+		if (error.code === "ENOENT") {
+			logger.error(
+				`\tYour ${optionName} "${path}" is not a valid directory on the filesystem.\n`
+			);
+		} else {
+			logger.error(
+				`\tYour ${optionName} "${path}" has invalid permissions.`
+			);
+		}
+	}
+	return false;
+}
+
+/**
+ * verifies the config paths provided against the filesystem
+ * @returns true (if paths are valid)
+ */
+async function checkConfigPaths(): Promise<void> {
+	const { action, linkDir, dataDirs, torrentDir, outputDir } =
+		getRuntimeConfig();
+	let pathFailure: number = 0;
+
+	if (!(await verifyPath(torrentDir, "torrentDir", constants.R_OK))) {
+		pathFailure++;
+	}
 
-function validateOptions() {
-	const {
-		action,
-		rtorrentRpcUrl,
-		qbittorrentUrl,
-		transmissionRpcUrl,
-		delugeRpcUrl,
-		dataDirs,
-		linkDir,
-		matchMode,
-		skipRecheck,
-	} = getRuntimeConfig();
 	if (
-		action === "inject" &&
-		!(
-			rtorrentRpcUrl ||
-			qbittorrentUrl ||
-			transmissionRpcUrl ||
-			delugeRpcUrl
-		)
+		action == Action.SAVE &&
+		!(await verifyPath(
+			outputDir,
+			"outputDir",
+			constants.R_OK | constants.W_OK
+		))
 	) {
-		throw new CrossSeedError(
-			"You need to specify --rtorrent-rpc-url, --transmission-rpc-url, --qbittorrent-url, or --deluge-rpc-url when using '-A inject'."
-		);
+		pathFailure++;
 	}
-	if ((dataDirs && !linkDir) || (!dataDirs && linkDir)) {
-		throw new CrossSeedError(
-			"Data based matching requires both --link-dir and --data-dirs"
-		);
+
+	if (
+		linkDir &&
+		!(await verifyPath(linkDir, "linkDir", constants.R_OK | constants.W_OK))
+	) {
+		pathFailure++;
+	}
+	if (dataDirs) {
+		for (const dataDir of dataDirs) {
+			if (!(await verifyPath(dataDir, "dataDirs", constants.R_OK))) {
+				pathFailure++;
+			}
+		}
 	}
-	if (matchMode == MatchMode.RISKY && skipRecheck) {
-		logger.warn(
-			"It is strongly recommended to not skip rechecking for risky matching mode"
+	if (pathFailure) {
+		throw new CrossSeedError(
+			`\tYour configuration is invalid, please see the ${
+				pathFailure > 1 ? "errors" : "error"
+			} above for details.`
 		);
 	}
 }
 
 export async function doStartupValidation(): Promise<void> {
-	logger.info("Validating your configuration...");
-	validateOptions();
+	const runtimeConfig: RuntimeConfig = getRuntimeConfig();
+	await checkConfigPaths();
+	setRuntimeConfig(runtimeConfig);
```

Comment:
this get-and-set thing seems redundant, you might have missed it during some refactoring

Reply:
At a certain point, for some reason, I was trying to sanitize the paths in dataDirs. I can't remember why I thought it was a good idea - and this got left over.

Obviously, we are terminating on errors. I think I might have been drunk when I was doing this as it was on NYE.

Reply:
Do you think I should pass the runtimeConfig to it through the function or use getRuntimeConfig() in doStartupValidation?